pkg/cable/wireguard/WGdriver.go

```diff
@@ -0,0 +1,365 @@
+	}
+
+	// handle public key
+	var remoteKey wgtypes.Key
+	var key string
+	if key, found = remoteEndpoint.Spec.BackendConfig[PublicKey]; !found {
+		return "", fmt.Errorf("missing peer public key")
+	}
+	if remoteKey, err = wgtypes.ParseKey(key); err != nil {
+		return "", fmt.Errorf("failed to parse public key %s: %v", key, err)
+	}
+	klog.V(log.TRACE).Infof("Connecting cluster %s endpoint %s with publicKey %s", remoteEndpoint.Spec.ClusterID, remoteIP.String(), remoteKey.String())
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+	var oldKey wgtypes.Key
+	if oldKey, found = w.peers[remoteEndpoint.Spec.ClusterID]; found {
+		if oldKey.String() == remoteKey.String() {
+			//TODO check that peer config has not changed (eg allowedIPs)
+			klog.V(log.TRACE).Infof("Skipping update of existing peer key %s", oldKey.String())
+			return ip, nil
+		}
+		// remove old
+		klog.V(log.TRACE).Infof("Removing old key %s", oldKey.String())
+		peerCfg := []wgtypes.PeerConfig{
+			{
+				PublicKey: remoteKey,
+				Remove:    true,
+			},
+		}
+		if err = w.client.ConfigureDevice(DefaultDeviceName, wgtypes.Config{
+			ReplacePeers: true,
+			Peers:        peerCfg,
+		}); err != nil {
+			klog.Errorf("Failed to remove old key %s: %v", oldKey.String(), err)
+		}
+		delete(w.peers, remoteEndpoint.Spec.ClusterID)
+		klog.V(log.TRACE).Infof("Successfully removed old key %s", oldKey.String())
+	}
+
+	// Set peer subnets
+	allowedIPs := make([]net.IPNet, len(remoteEndpoint.Spec.Subnets))
+	var cidr *net.IPNet
+	for i, sn := range remoteEndpoint.Spec.Subnets {
+		if _, cidr, err = net.ParseCIDR(sn); err != nil {
+			return "", fmt.Errorf("failed to parse subnet %s: %v", sn, err)
+		}
+		allowedIPs[i] = *cidr
+	}
+
+	// configure peer
+	peerCfg := []wgtypes.PeerConfig{{
+		PublicKey:    remoteKey,
+		Remove:       false,
+		UpdateOnly:   false,
+		PresharedKey: nil,
+		Endpoint: &net.UDPAddr{
+			IP:   remoteIP,
+			Port: DefaultListenPort,
+		},
+		PersistentKeepaliveInterval: nil,
+		ReplaceAllowedIPs:           true,
+		AllowedIPs:                  allowedIPs,
+	}}
+	if err = w.client.ConfigureDevice(DefaultDeviceName, wgtypes.Config{
+		ReplacePeers: false,
+		Peers:        peerCfg,
+	}); err != nil {
+		return "", fmt.Errorf("failed to configure peer: %v", err)
+	}
+	// verify peer was added
+	var d *wgtypes.Device
+	if d, err = w.client.Device(DefaultDeviceName); err != nil {
+		klog.Errorf("Failed to find wireguard device by name: %v", err)
+	} else {
+		found = false
+		for _, p := range d.Peers {
+			if p.PublicKey.String() == remoteKey.String() {
+				found = true
+				klog.V(log.TRACE).Infof("Peer configured: %+v", p)
+				break
+			}
+		}
+		if !found {
+			klog.Errorf("Failed to verify peer configuration")
+		}
+	}
+	w.peers[remoteEndpoint.Spec.ClusterID] = remoteKey
+
+	// Add routes to peer
+	//TODO save old routes for removal
+	var wg netlink.Link
+	if wg, err = netlink.LinkByName(DefaultDeviceName); err != nil {
```

Comment:
Can't use use `w.link` instead of retrieving it again?

Reply:
done